File: sjwt_token.js
const jwt = require('jsonwebtoken');
const crypto = require('crypto');

const jwtSecretKey = 'your_secret_key';

function generateToken(user) {
  const header = base64UrlEncode(JSON.stringify({
    alg: 'HS256',
    typ: 'JWT'
  }));

  const payload = base64UrlEncode(JSON.stringify({
    iss: 'senad_cavkusic',
    aud: '.sarajevoweb.com',
    iat: Math.floor(Date.now() / 1000),
    exp: Math.floor(Date.now() / 1000) + 3600,
    idv: user.id,
    ipv: getUniqueBrowserId()
  }));

  const dataToSign = `${header}.${payload}`;
  const signature = base64UrlEncode(
    crypto
      .createHmac('sha256', jwtSecretKey)
      .update(dataToSign)
      .digest('binary')
  );

  return `${dataToSign}.${signature}`;
}

function getUniqueBrowserId() {
  const userAgent = req.headers['user-agent'];
  const ipAddress =
    req.headers['x-forwarded-for'] ||
    req.connection.remoteAddress ||
    req.socket.remoteAddress ||
    req.connection.socket.remoteAddress;

  const ipWithoutDots = ipAddress.replace(/\./g, '');
  const uniqueId = crypto.createHash('sha256').update(userAgent + ipWithoutDots).digest('hex');

  return uniqueId;
}

function validateToken(token) {
  const tokenParts = token.split('.');
  if (tokenParts.length !== 3) {
    throw new Error('Invalid token format');
  }

  const headerBase64 = tokenParts[0];
  const payloadBase64 = tokenParts[1];
  const signatureBase64 = tokenParts[2];

  const header = JSON.parse(base64UrlDecode(headerBase64));
  const payload = JSON.parse(base64UrlDecode(payloadBase64));
  const signature = base64UrlDecode(signatureBase64);

  if (!header.alg || header.alg !== 'HS256') {
    throw new Error('Unexpected or missing algorithm in token header');
  }

  const expectedSignature = crypto
    .createHmac('sha256', jwtSecretKey)
    .update(`${headerBase64}.${payloadBase64}`)
    .digest('binary');

  if (!crypto.timingSafeEqual(Buffer.from(expectedSignature, 'binary'), Buffer.from(signature, 'binary'))) {
    throw new Error('Invalid token signature');
  }

  if (!payload.iss || payload.iss !== 'senad_cavkusic') {
    throw new Error('Invalid token issuer');
  }

  if (payload.exp && payload.exp < Math.floor(Date.now() / 1000)) {
    throw new Error('Token has expired');
  }

  const userId = payload.idv || null;
  const userIP = getUniqueBrowserId();
  const storedIP = payload.ipv || null;

  if (storedIP !== userIP) {
    throw new Error('Invalid IP address');
  }

  return userId;
}

function base64UrlEncode(data) {
  let encoded = Buffer.from(data).toString('base64');
  encoded = encoded.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
  return encoded;
}

function base64UrlDecode(data) {
  const paddingLength = (4 - (data.length % 4)) % 4;
  const paddedData = data + '='.repeat(paddingLength);
  return Buffer.from(paddedData, 'base64').toString();
}

module.exports = {
  generateToken,
  validateToken
};

File: sjwt_token.py
import hashlib
import base64
import json
import hmac
import time

jwtSecretKey = "your_jwt_secret_key"

def generate_token(user):
    header = base64_url_encode(json.dumps({
        'alg': 'HS256',
        'typ': 'JWT'
    }))

    payload = base64_url_encode(json.dumps({
        'iss': 'senad_cavkusic',
        'aud': '.sarajevoweb.com',
        'iat': int(time.time()),
        'exp': int(time.time()) + 3600,
        'idv': user['id'],
        'ipv': grab_user_ip_address()
    }))

    data_to_sign = f"{header}.{payload}"
    signature = base64_url_encode(hmac.new(jwtSecretKey.encode(), data_to_sign.encode(), hashlib.sha256).digest())

    return f"{data_to_sign}.{signature}"

def grab_user_ip_address():
    # Retrieve user IP address based on your server environment
    # Implement the logic to get the IP address here
    return user_ip_address

def validate_token(token):
    token_parts = token.split('.')
    if len(token_parts) != 3:
        raise Exception('Invalid token format')

    header_base64 = token_parts[0]
    payload_base64 = token_parts[1]
    signature_base64 = token_parts[2]

    header = json.loads(base64_url_decode(header_base64))
    payload = json.loads(base64_url_decode(payload_base64))
    signature = base64_url_decode(signature_base64)

    # Verify the header
    if 'alg' not in header or header['alg'] != 'HS256':
        raise Exception('Unexpected or missing algorithm in token header')

    # Verify the signature
    expected_signature = hmac.new(jwtSecretKey.encode(), f"{header_base64}.{payload_base64}".encode(), hashlib.sha256).digest()
    if not hmac.compare_digest(expected_signature, signature):
        raise Exception('Invalid token signature')

    # Verify the issuer
    if 'iss' not in payload or payload['iss'] != 'senad_cavkusic':
        raise Exception('Invalid token issuer')

    # Verify the token hasn't expired
    if 'exp' in payload and payload['exp'] < int(time.time()):
        raise Exception('Token has expired')

    # Verify the IP address
    user_id = payload.get('idv')
    user_ip = grab_user_ip_address()
    stored_ip = payload.get('ipv')

    if stored_ip != user_ip:
        raise Exception('Invalid IP address')

    return user_id

def base64_url_encode(data):
    encoded_bytes = base64.urlsafe_b64encode(data.encode()).rstrip(b'=')
    return encoded_bytes.decode()

def base64_url_decode(data):
    padding = len(data) % 4
    padded_data = data + '=' * (4 - padding)
    decoded_bytes = base64.urlsafe_b64decode(padded_data.encode())
    return decoded_bytes.decode()
